### ISSUE TITLE
Handling recompile in apply rotary emb

### ISSUE THREAD

**Issue by conceptofmind** [Dao-AILab/quack#150]
Hi all,

I was wondering what the recommended way to handle recompilation would be in the `apply_rotary_emb` function.

Given that during varlen training, the max seqlen in the flattened batch is always different, this causes a recompilation with `model = torch.compile(model, dynamic=False, fullgraph=True,)`. Many data collators or loaders do `max_seqlen = lengths.max().item()`.
```
V0610 23:38:24.321000 168840 .venv/lib/python3.12/site-packages/torch/_dynamo/guards.py:4760] [0/2] [__recompiles]     triggered by the following guard failure(s):
V0610 23:38:24.321000 168840 .venv/lib/python3.12/site-packages/torch/_dynamo/guards.py:4760] [0/2] [__recompiles]     - 0/1: max_length_k == 496                                      # return ApplyRotaryEmb.apply(  # quack/rotary.py:667 in apply_rotary_emb
```

Would it make sense to bucket the max seqlen to the total context length that is used for training, such as 512? Or would it be better to just not do fullgraph, introduce a graph break, and to mark the apply rotary emb function with something such as `dynamic=True`.

Just trying to think about the best way to approach and handle this in quack.

Thank you,

Enrico

**Comment by tridao** [comment]
are you referring to torch compile or cute compile (compiling the kernel)? The kernel compilation doesn't depend on the seqlen, only on the dtype and dim.
https://github.com/Dao-AILab/quack/blob/99bd7973bf3dc6db40961e413d4bdfea6c6fee3e/quack/rotary.py#L290
If you're talking about torch compile, then it seems like a general issue with torch compile and dynamic shape, and not particular to rotary?

**Comment by conceptofmind** [comment]
> are you referring to torch compile or cute compile (compiling the kernel)? The kernel compilation doesn't depend on the seqlen, only on the dtype and dim.
> 
> [quack/quack/rotary.py](https://github.com/Dao-AILab/quack/blob/99bd7973bf3dc6db40961e413d4bdfea6c6fee3e/quack/rotary.py#L290)
> 
> Line 290 in [99bd797](/Dao-AILab/quack/commit/99bd7973bf3dc6db40961e413d4bdfea6c6fee3e)
> 
>  def compile( 
> 
> If you're talking about torch compile, then it seems like a general issue with torch compile and dynamic shape, and not particular to rotary?

Actually, it is probably good to specify that now.

Since I am passing max seq len:
```python
        q = apply_rotary_emb(
            q,
            cos,
            sin,
            cu_seqlens=cu_seqlens_q,
            max_seqlen=max_seqlen_q,
            inplace=True,
        )
        k = apply_rotary_emb(
            k,
            cos,
            sin,
            cu_seqlens=cu_seqlens_k,
            max_seqlen=max_seqlen_k,
            inplace=True,
        )
```
This is likely then a `torch.compile()` issue, and I will need to introduce a graph break with dynamic shaping to resolve it.
```
V0611 03:04:16.327000 182716 .venv/lib/python3.12/site-packages/torch/_dynamo/guards.py:4760] [0/1] [__recompiles] Recompiling function forward in /e/project1/reformo/enrico/BlitzBERT/src/blitzbert/model/modeling_blitzbert_fa4.py:314
V0611 03:04:16.327000 182716 .venv/lib/python3.12/site-packages/torch/_dynamo/guards.py:4760] [0/1] [__recompiles]     triggered by the following guard failure(s):
V0611 03:04:16.327000 182716 .venv/lib/python3.12/site-packages/torch/_dynamo/guards.py:4760] [0/1] [__recompiles]     - 0/0: max_length_k == 3765                                     # return ApplyRotaryEmb.apply(  # quack/rotary.py:667 in apply_rotary_emb
V0611 03:04:16.327000 182716 .venv/lib/python3.12/site-packages/torch/_dynamo/guards.py:4760] [0/1] [__recompiles]     - User stack trace:
V0611 03:04:16.327000 182716 .venv/lib/python3.12/site-packages/torch/_dynamo/guards.py:4760] [0/1] [__recompiles]     -   File "/e/project1/reformo/enrico/BlitzBERT/src/blitzbert/model/modeling_blitzbert_fa4.py", line 323, in forward
V0611 03:04:16.327000 182716 .venv/lib/python3.12/site-packages/torch/_dynamo/guards.py:4760] [0/1] [__recompiles]     -     output = self.model(
V0611 03:04:16.327000 182716 .venv/lib/python3.12/site-packages/torch/_dynamo/guards.py:4760] [0/1] [__recompiles]     -   File "/e/project1/reformo/enrico/BlitzBERT/src/blitzbert/model/modeling_blitzbert_fa4.py", line 266, in forward
V0611 03:04:16.327000 182716 .venv/lib/python3.12/site-packages/torch/_dynamo/guards.py:4760] [0/1] [__recompiles]     -     return self.transformer(
V0611 03:04:16.327000 182716 .venv/lib/python3.12/site-packages/torch/_dynamo/guards.py:4760] [0/1] [__recompiles]     -   File "/e/project1/reformo/enrico/BlitzBERT/src/blitzbert/model/modeling_blitzbert_fa4.py", line 183, in forward
V0611 03:04:16.327000 182716 .venv/lib/python3.12/site-packages/torch/_dynamo/guards.py:4760] [0/1] [__recompiles]     -     hidden_states = layer(
V0611 03:04:16.327000 182716 .venv/lib/python3.12/site-packages/torch/_dynamo/guards.py:4760] [0/1] [__recompiles]     -   File "/e/project1/reformo/enrico/BlitzBERT/src/blitzbert/model/modeling_blitzbert_fa4.py", line 133, in forward
V0611 03:04:16.327000 182716 .venv/lib/python3.12/site-packages/torch/_dynamo/guards.py:4760] [0/1] [__recompiles]     -     hidden_states = hidden_states + self.attn(
V0611 03:04:16.327000 182716 .venv/lib/python3.12/site-packages/torch/_dynamo/guards.py:4760] [0/1] [__recompiles]     -   File "/e/project1/reformo/enrico/BlitzBERT/src/blitzbert/model/modeling_blitzbert_fa4.py", line 80, in forward
V0611 03:04:16.327000 182716 .venv/lib/python3.12/site-packages/torch/_dynamo/guards.py:4760] [0/1] [__recompiles]     -     k = apply_rotary_emb(
V0611 03:04:16.327000 182716 .venv/lib/python3.12/site-packages/torch/_dynamo/guards.py:4760] [0/1] [__recompiles]     -   File "/e/project1/reformo/enrico/BlitzBERT/.venv/lib/python3.12/site-packages/quack/rotary.py", line 667, in apply_rotary_emb
V0611 03:04:16.327000 182716 .venv/lib/python3.12/site-packages/torch/_dynamo/guards.py:4760] [0/1] [__recompiles]     -     return ApplyRotaryEmb.apply(
 ```
If so, happy to close out this issue.

Thank you,

Enrico

**Comment by tridao** [comment]
if there's a different way to register the rotary with torch library that would avoid this torch recompile, we're happy to do it. But i'm not an expert on torch compile

**Comment by conceptofmind** [comment]
> if there's a different way to register the rotary with torch library that would avoid this torch recompile, we're happy to do it. But i'm not an expert on torch compile

I will open an issue on PyTorch and see if I can get some input to determine why this occurs.

I will close out this issue for now and reopen once I have more information.

Thanks!
